### PR TITLE
[VPlan] Add commutative binary OR matcher, use in transform.

### PR DESCRIPTION
**Pull request by fhahn**
Split off from https://github.com/llvm/llvm-project/pull/89386, this
extends the binary matcher to support matching commuative operations.
This is used for a new m_c_BinaryOr matcher, used in simplifyRecipe.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 1]
```diff
@@ -935,6 +935,22 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
...
+  if (match(&R,
+            m_c_BinaryOr(m_LogicalAnd(m_VPValue(X), m_VPValue(Y)),
+                         m_LogicalAnd(m_VPValue(X1), m_Not(m_VPValue(Y1))))) &&
+      X == X1 && Y == Y1) {
+    R.getVPSingleValue()->replaceAllUsesWith(X);
+    return;
+  }
+
   if (match(&R, m_CombineOr(m_Mul(m_VPValue(A), m_SpecificInt(1)),
                             m_Mul(m_SpecificInt(1), m_VPValue(A)))))
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:955]: Could this be simplified by defining m_Mul as commutative? Possibly in a separate patch.

File: llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h [hunk 2]
```diff
@@ -268,10 +274,23 @@ m_Mul(const Op0_t &Op0, const Op1_t &Op1) {
   return m_Binary<Instruction::Mul, Op0_t, Op1_t>(Op0, Op1);
 }
 
+template <typename Op0_t, typename Op1_t, bool Commutative = false>
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:277]: Should the default for Or be Commutative = true?

File: llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h [hunk 3]
```diff
@@ -268,10 +274,23 @@ m_Mul(const Op0_t &Op0, const Op1_t &Op1) {
...
+inline AllBinaryRecipe_match<Op0_t, Op1_t, Instruction::Or, Commutative>
+m_BinaryOr(const Op0_t &Op0, const Op1_t &Op1) {
+  return m_Binary<Instruction::Or, Op0_t, Op1_t, Commutative>(Op0, Op1);
+}
+
+template <typename Op0_t, typename Op1_t>
+inline AllBinaryRecipe_match<Op0_t, Op1_t, Instruction::Or,
+                             /*Commutative*/ true>
+m_c_BinaryOr(const Op0_t &Op0, const Op1_t &Op1) {
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:286]: "m_c_" stands for commutative?

**Review by ayalz — COMMENTED** [review]
Thanks for splitting! Curious to see which tests are still affected after rebase, whether additional tests may be needed.

File: llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:277]: The default (=false) is in line with the IR based pattern matcher I think. It only has the argument to allow for a slightly simpler m_c_BinaryOr implementation.

File: llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h (hunk 3 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:286]: Yes, this is like in the IR based pattern matcher.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Yes!

**Review by fhahn — COMMENTED** [review]
Just updated to latest main, including #89386

File: llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h (hunk 2 above)
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:277]: Ok, worth a comment, as conceptually matching Or's should be commutative by default, so unclear when one would use the non commutative version, which could still be provided albeit as opt in.

File: llvm/test/Transforms/LoopVectorize/AArch64/masked-call.ll [hunk 4]
```diff
@@ -402,10 +402,9 @@ define void @test_widen_if_then_else(ptr noalias %a, ptr readnone %b) #4 {
 ; TFCOMMON-NEXT:    [[TMP11:%.*]] = call <vscale x 2 x i64> @foo_vector(<vscale x 2 x i64> zeroinitializer, <vscale x 2 x i1> [[TMP10]])
 ; TFCOMMON-NEXT:    [[TMP12:%.*]] = select <vscale x 2 x i1> [[ACTIVE_LANE_MASK]], <vscale x 2 x i1> [[TMP8]], <vscale x 2 x i1> zeroinitializer
 ; TFCOMMON-NEXT:    [[TMP13:%.*]] = call <vscale x 2 x i64> @foo_vector(<vscale x 2 x i64> [[WIDE_MASKED_LOAD]], <vscale x 2 x i1> [[TMP12]])
-; TFCOMMON-NEXT:    [[TMP14:%.*]] = or <vscale x 2 x i1> [[TMP10]], [[TMP12]]
```
Comment (ayalz) [inline llvm/test/Transforms/LoopVectorize/AArch64/masked-call.ll:405]: This and below match X&&!Y || X&&Y --> X, exercising commutativity.

**Review by ayalz — APPROVED** [review]
LGTM, exercised by a handful of tests.

File: llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:281]: Added in the latest version, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h [hunk 5]
```diff
@@ -268,10 +274,21 @@ m_Mul(const Op0_t &Op0, const Op1_t &Op1) {
   return m_Binary<Instruction::Mul, Op0_t, Op1_t>(Op0, Op1);
 }
 
-template <typename Op0_t, typename Op1_t>
-inline AllBinaryRecipe_match<Op0_t, Op1_t, Instruction::Or>
+/// Match a binary OR operation. Note that while conceptually the perands can
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:277]:
```suggestion
/// Match a binary OR operation. Note that while conceptually the operands can
```

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:277]: Ah yes, should be fixed, thanks!

**Pull request merged**